ibis/tests/all/test_vectorized_udf.py

```diff
@@ -74,13 +74,83 @@ def foo3(v, **kwargs):
+def test_valid_args(backend, alltypes, df):
+    # Test UDF with variable-length positional arguments (e.g. *args)
+
+    @elementwise(input_type=[dt.double, dt.string], output_type=dt.double)
+    def foo1(*args):
+        return args[0] + len(args[1])
+
+    result = alltypes.mutate(
+        v1=foo1(alltypes['double_col'], alltypes['string_col']),
+    ).execute()
+
+    expected = df.assign(v1=df['double_col'] + len(df['string_col']),)
+
+    backend.assert_frame_equal(result, expected)
+
+
+@pytest.mark.only_on_backends([Pandas, PySpark])
+@pytest.mark.xfail_unsupported
+def test_valid_args_and_kwargs(backend, alltypes, df):
+    # Test UDFs with both variable-length positional arguments (e.g. *args)
+    # and keyword arguments
+
+    @elementwise(input_type=[dt.double, dt.string], output_type=dt.double)
+    def foo1(*args, amount):
+        # UDF with *args and a keyword-only argument
+        return args[0] + len(args[1]) + amount
+
+    @elementwise(input_type=[dt.double, dt.string], output_type=dt.double)
+    def foo2(*args, **kwargs):
+        # UDF with *args and **kwargs
+        return args[0] + len(args[1]) + kwargs.get('amount', 1)
+
+    result = alltypes.mutate(
+        v1=foo1(alltypes['double_col'], alltypes['string_col'], amount=1),
+        v2=foo1(alltypes['double_col'], alltypes['string_col'], amount=2),
+        v3=foo2(alltypes['double_col'], alltypes['string_col']),
+        v4=foo2(alltypes['double_col'], alltypes['string_col'], amount=2),
+        v5=foo2(alltypes['double_col'], alltypes['string_col'], amount=3),
+    ).execute()
+
+    expected = df.assign(
+        v1=df['double_col'] + len(df['string_col']) + 1,
+        v2=df['double_col'] + len(df['string_col']) + 2,
+        v3=df['double_col'] + len(df['string_col']) + 1,
+        v4=df['double_col'] + len(df['string_col']) + 2,
+        v5=df['double_col'] + len(df['string_col']) + 3,
+    )
+
+    backend.assert_frame_equal(result, expected)
+
+
 @pytest.mark.only_on_backends([Pandas, PySpark])
 @pytest.mark.xfail_unsupported
 def test_invalid_kwargs(backend, alltypes):
-    # Test different invalid UDFs
+    # Test that defining a UDF with a non-column argument that is not a
+    # keyword argument raises an error
 
     with pytest.raises(TypeError, match=".*must be defined as keyword only.*"):
 
         @elementwise(input_type=[dt.double], output_type=dt.double)
         def foo1(v, amount):
             return v + 1
+
+
+@pytest.mark.only_on_backends([Pandas, PySpark])
+@pytest.mark.xfail_unsupported
+def test_invalid_args(backend, alltypes, df):
+    # Test that defining a UDF with both positional arguments and variable-
+    # length positional arguments (e.g. *args) raises an error
+
+    with pytest.raises(
+        com.IbisError,
+        match=".*cannot have both positional arguments and \\*args",
+    ):
+
+        @elementwise(input_type=[dt.double, dt.string], output_type=dt.double)
+        def foo1(v, *args):
```

Comment:
It's not clear why this is disallowed. Can you explain?

Reply:
This was to limit users to only using one type of positional argument per UDF function signature (either explicit positional arguments or `*args`) so that the function signatures are a little clearer

```
@elementwise(input_type=[dt.double, dt.double], output_type=dt.double)
def add_cols(col1, col2):
    return col1 + col2
```
or
```
@elementwise(input_type=[dt.double, dt.double], output_type=dt.double)
def add_cols(*cols):
    return col[0] + col[1]
```

versus

```
# This is being disallowed
@elementwise(input_type=[dt.double, dt.double], output_type=dt.double)
def add_cols(col1, *addtl_cols):
    return col1 + addtl_cols[0]
```

Reply:
I think we should not add addition constraint to ibis UDF because it is a "confusing way to write Python code". I agree that "add_cols(col1, *addtl_cols)" is more confusing than "add_cols(col1, col2)" and "add_cols(*cols)". However, the confusion doesn't come from anything ibis specific - it's confusing because it's a bad way of using Python for this use case. And I don't think we should try to disallow people to write "bad Python code" in ibis. Ibis shouldn't have opinion on that I think.

Reply:
I think that's a fair point, having the check is extra complexity on our side to catch Python syntax that is debatably confusing but legal

Reply:
I removed the check